### PR TITLE
[wptrunner] Run larger groups first

### PR DESCRIPTION
**Pull request by jonathan-j-lee**
This change orders groups within each test type in the test queue by
decreasing total timeout. This heuristic attempts to distribute work to
test runners more evenly by scheduling small groups around larger ones.

A large group that runs near the end of the suite cannot be split, so
some runners may become idle while waiting for stragglers. For example,
there's a 1 min gap between the first and last idle processes in [this
initial run](https://chromium-swarm.appspot.com/task?id=62b302bbdd0bbf10) (indicated by `No more tests` messages).

Empirically, in Chromium CI, this change reduces total test runtime
meaningfully:
* Chrome on Linux: [~120 min](https://ci.chromium.org/ui/p/chromium/builders/ci/linux-wpt-fyi-rel/42930/overview) &rightarrow; [~115 min](https://ci.chromium.org/ui/p/chromium/builders/try/linux-wpt-fyi-rel/2035/overview)
* Content Shell on Linux: [~100 min](https://ci.chromium.org/ui/p/chromium/builders/ci/linux-wpt-content-shell-fyi-rel/9981/overview) &rightarrow; [~90 min](https://ci.chromium.org/ui/p/chromium/builders/try/linux-wpt-content-shell-fyi-rel/1850/overview)

### CODE REVIEW

**Review by jgraham — COMMENTED** [review]
So generally I think this is OK (and I like the refactoring).

Just a couple of questions:
* Should we skip this for `GroupFileTestSource`? In theory one could use that to provide an explicit ordering using out-of-band information (e.g. actual measured group duration), so the timeout based measurement might be a deoptimisation (and surprising).
* Should we skip for the case where we're only running one process? In that case consistent ordering might be easier to understand and there isn't any perf win.

**Comment by jonathan-j-lee** [comment]
> Should we skip this for `GroupFileTestSource`?

Maybe I'm misunderstanding, but it seems like the `--test-groups` format `{group: [test_id, ...]}` cannot specify group order? (If it could, I probably would have done that instead of making this change.)

https://github.com/web-platform-tests/wpt/blob/c2ec4c5131d43e0a8b141afa5acbd5bd36181341/tools/wptrunner/wptrunner/testloader.py#L539-L545

* The order of test types within the queue is determined by `tests_by_type`'s iteration order (if any), which is controlled by the `make_queue()` caller.
* The order of groups within a test type's section is determined by `tests_by_group`'s iteration order, but that's a `defaultdict` that is inserted into by `tests_by_type`'s values mapped through `group_by_test`.

In short, the existing group order for `--test-groups` doesn't seem well-defined beyond "test types are contiguous".

> Should we skip for the case where we're only running one process?

Done

**Comment by WeizhongX** [comment]
With the data you listed, sounds like wptrunner side uses less time than RWT?

**Comment by WeizhongX** [comment]
No, I have a CL that excludes virtual tests from RWT side also. And the total test time is about 100m.

**Comment by jonathan-j-lee** [comment]
I don't follow? ~100 min is the current baseline that the CI `linux-wpt-content-shell-fyi-rel` builds on main show; if your CL shows that, then it's likely it didn't have a large impact.

I gathered my data with multiple try builds in [this change](https://chromium-review.googlesource.com/c/chromium/src/+/4575972?tab=checks), which show reductions in max shard runtimes and runtimes for the `not-site-per-process` suite as well.

**Comment by jonathan-j-lee** [comment]
> No, I have a CL that excludes virtual tests from RWT side also. And the total test time is about 100m.

Are you sure it wasn't just a fluke? `blink_wpt_tests` for `Linux Tests` shows [~100 min](https://ci.chromium.org/ui/p/chromium/builders/ci/Linux%20Tests/131500/overview) with virtual tests.

The try builds I gathered my data from are in [this change](https://chromium-review.googlesource.com/c/chromium/src/+/4575972?tab=checks), which also reduced max shard runtimes and `not-site-per-process` runtimes.

**Review by WeizhongX — APPROVED** [review]
